lib/wsapp.js

```diff
@@ -12,15 +12,81 @@ var MessageBroker = require('./messagebroker');
 
 log.level = 'debug';
 
-var WsApp = function() {};
-
-WsApp.prototype._unauthorized = function(socket) {
-  socket.emit('unauthorized');
-  socket.disconnect();
+var WsApp = function() {
+    this.connectedWallets = {};
+    this.authorizedWallets = {};
 };
 
+function keysToArray(dict){
+  var ret = [];
+  for (var key in dict)
+    ret.push(key);
+  return ret;
+}
+
 WsApp.prototype._handleNotification = function(notification) {
-  this.io.to(notification.walletId).emit('notification', notification);
+  var namespace = this.io.to(notification.walletId);
+  var to = this.authorizedWallets[notification.walletId] || {};
+  console.log('Notification: ', notification);
+  namespace.emit('multimessage', {to: keysToArray(to), message: 'notification', payload: notification});
+};
+
+function _unauthorized(socket, connectionId) {
+  socket.emit('message', {to: connectionId, message: 'unauthorized'});
+};
+
+WsApp.prototype._walletConnected = function(socket, data){
+  if (!data || !data.from)
+    return;
+  var connectionId = data.from;
+  var wallet = this.connectedWallets[connectionId] = {
+    socket: socket,
+    nonce: Uuid.v4()
+  };
+  socket.emit('message', {to: connectionId, message: 'challenge', payload: wallet.nonce});
+};
+
+WsApp.prototype._authorizeWallet = function(socket, data){
+  if (!data || !data.from || !data.payload || !(data.from in this.connectedWallets))
+    return;
+  var connectionId = data.from;
+  if (!connectionId || !(connectionId in this.connectedWallets))
+    return;
+
+  var nonce = data.payload.message;
+  var connectedWallet = this.connectedWallets[connectionId];
+  var expectedNonce = connectedWallet.nonce;
+  if (nonce != expectedNonce){
+    _unauthorized(socket, connectionId);
+    return;
+  }
+  
+  var self = this;
+  
+  WalletService.getInstanceWithAuth(
+    data.payload,
+    function(err, service) {
+      if (err){
+        _unauthorized(socket, connectionId);
+        return;
+      }
+
+      socket.join(service.walletId);
+      var dict = null;
+      if (service.walletId in self.authorizedWallets)
+        dict = self.authorizedWallets[service.walletId];
+      else
+        dict = self.authorizedWallets[service.walletId] = {};
+      dict[connectionId] = null;
```

Comment:
I think this is a little obscure. Any reason to use a dictionary instead of just a regular array? Something like:
```javascript
self.authorizedWallets[service.walletId].push(connectionId);
```
would be simpler IMO.